content/blog/2019-10-23-register-allocation/index.md

```diff
@@ -0,0 +1,142 @@
++++
+title = "Register Allocation for Bril"
+extra.author = "Hongbo Zhang, Sachille Atapattu, Wen-Ding Li"
+extra.bio = """
+  Hongbo Zhang is a first PhD student in computer science. He is interested in systems and computer architectures. He is also okay archer shooting recurve bow.
+  [Wen-Ding Li](https://www.cs.cornell.edu/~wdli/) is a Ph.D. student at Cornell interested in high-performance code generation.
+"""
++++
+# Register Allocation for Bril
+
+Processors usually have registers which can be used to store variables for faster access. Program will run faster if we can put variables in these registers. Nevertheless, registers are limited, so often times we still need to access memory (or cache) instead of registers and memory access are costly compared with register access. Although modern program often use large number of variables, we can analyze their live range so the same registers may be able to hold different varialbes throughout the program. Then, we can put more varialbes in the registers which may lead to fewer memory access and get a faster program. 
+
+The goal of our project is to perform register allocation on Bril program and anaylze its performance. In this project, we implemented register allocation for Bril via graph coloring. Our project can be found in [this repository](https://github.com/xu3kev/bril) under `regalloc` directory.
+
+## Method
+Our approach for the register allocation can be divided in three main parts. First, we perform liveness analysis and then build a graph. Second, we solve approximate k graph coloring problem to obtain a mapping from each varialbe to corresponding register. At last, we generate Bril code based on the mapping.
+
+### Liveness Analysis
+We use data flow analysis to get the live varialbes at each instruction of the program. First we transform the input Bril program to control flow graph (cfg) and then perform liveness analysis with a standard backward flow anaylsis. Once we obtain the live varialbes in the end of each code block, we start from that and go backward: when we see a definition of a varialbe, we kill the variable, and when we see the use of the varialbe, we mark it as alive. Then, we can get the live varialbes at every instructions.
+
+### Graph Coloring
+With the liveness ranges of all local variables, we can know which variables are "alive" at the same time. If two variables have overlapped liveness ranges, it means that they cannot be allocated to the same register. Otherwise, they may be allocated to the same register. Our goal is to have register assignment so that we can allocate as many variables on registers as possible.
+We solve this question by graph coloring. Each node in the graph represents a local variable in the function. There is an edge between two nodes if the two variables have overlapped liveness range. We just needs to compute a color assignment with at most *k* colors, such that every pair of nodes on the two sides of an edge have different colors.
```

Comment:
Some more example of plurality corrections:

variable -> variables
range -> ranges
needs -> need